Fix issue #966 (initializing class instances twice)

kinke opened the pull request:
We initialize the instance in `DtoNewClass()`, so elide pre-initializing it in `_d_newclass()`.

kinke committed:
Fix issue #966 (initializing class instances twice)

We initialize the instance in DtoNewClass(), so elide pre-initializing it
in _d_newclass().

PetarKirov commented:
Can you upstream the change?

kinke commented:
What do you mean? Does DMD have a similar post-initialize step rendering the first init in _d_newclass() superfluous?

dnadlinger commented:
Remind me – why can't we just go with the DMD semantics instead?

kinke commented:
@klickverbot: Back then, I couldn't simply get rid of our second initialization due to the `GarbageCollect2Stack` pass, which simply replaces the `_d_newclass()` call with an alloca (hence missing the initialization). So either we remove the 1st init in druntime, or we remove the 2nd initialization in DtoNewClass() but add a manual init in GC2Stack.
The DtoNewClass() init provides two benefits:
1. It's done at the `_d_newclass()` call site, i.e., the init is inlined so to speak.
2. The memcpy src is the global init symbol constant directly, while `_d_newclass()` goes the `ClassInfo.init.ptr` route.

dnadlinger commented:
I don't think 2) is a particularly strong argument here, since `_d_newclass` already contains a bunch of `ClassInfo` lookups and a call into the GC. 1) might be relevant in some situations, though.

In any case, I I think changing the _d_newclass semantics might be a bit dangerous w.r.t. existing code that uses the druntime functions directly. While we of course never made a stability guarantee for these functions, it seems like there is quite the potential for breakage:

https://github.com/msgpack/msgpack-d/pull/65/files
https://github.com/search?l=d&q=_d_newclass&ref=reposearch&type=Code&utf8=%E2%9C%93

We could always create a second runtime function that skips the initialisation, though – like the old _d_allocclass.

kinke commented:
> like the old _d_allocclass

Alright, that doesn't sad too bad. How'd you prefer it? Rename the existing `_d_newclass()` and take another `bool initialize` parameter, then add 2 LDC specific `_d_newclass()/_d_allocclass()` calling this function (to keep the diff small), or adding a near 1:1 copy of `_d_newclass()` (to avoid another nested call for debug builds)?

kinke committed:
Keep _d_newclass() semantics, re-introduce _d_allocclass() instead

To keep 3rd party code from breaking. LDC will use _d_allocclass() in
DtoNewClass() and take care of initialization.

kinke commented:
Updated accordingly.
